crates/bevy_ecs/src/system/observer_system.rs

```diff
@@ -0,0 +1,71 @@
+use bevy_utils::all_tuples;
+
+use crate::{
+    prelude::{Bundle, Observer},
+    system::{System, SystemParam, SystemParamFunction, SystemParamItem},
+};
+
+use super::IntoSystem;
+
+/// Implemented for systems that have an [`Observer`] as the first argument.
+pub trait ObserverSystem<E: 'static, B: Bundle>:
+    System<In = Observer<'static, E, B>, Out = ()> + Send + 'static
+{
+}
+
+impl<E: 'static, B: Bundle, T: System<In = Observer<'static, E, B>, Out = ()> + Send + 'static>
+    ObserverSystem<E, B> for T
+{
+}
+
+/// Implemented for systems that convert into [`ObserverSystem`].
+pub trait IntoObserverSystem<E: 'static, B: Bundle, M>: Send + 'static {
+    /// The type of [`System`] that this instance converts into.
+    type System: ObserverSystem<E, B>;
+
+    /// Turns this value into its corresponding [`System`].
+    fn into_system(this: Self) -> Self::System;
+}
+
+impl<S: IntoSystem<Observer<'static, E, B>, (), M> + Send + 'static, M, E: 'static, B: Bundle>
+    IntoObserverSystem<E, B, M> for S
+where
+    S::System: ObserverSystem<E, B>,
+{
+    type System = <S as IntoSystem<Observer<'static, E, B>, (), M>>::System;
+
+    fn into_system(this: Self) -> Self::System {
+        IntoSystem::into_system(this)
+    }
+}
+
+macro_rules! impl_system_function {
+    ($($param: ident),*) => {
+        #[allow(non_snake_case)]
+        impl<E: 'static, B: Bundle, Func: Send + Sync + 'static, $($param: SystemParam),*> SystemParamFunction<fn(Observer<E, B>, $($param,)*)> for Func
+        where
+        for <'a> &'a mut Func:
+                FnMut(Observer<E, B>, $($param),*) +
+                FnMut(Observer<E, B>, $(SystemParamItem<$param>),*)
+        {
+            type In = Observer<'static, E, B>;
```

Comment:
This feels like a slight mismatch conceptually with the current `System<In = X>` impls, which expect `In<X>`, not `X`.

That being said, it seems like this pattern should be generalizable ... we shouldn't need to re-implement this trait for each different input type, especially given the compile time cost of generating all of these impls.

We should explore implementing this pattern in a more general way.

Reply:
Arg actually probably not possible given that using a `T` generic creates overlaps with the "normal" impls. Thats what justified the `In` wrapper in the first place / I remember wrestling with this problem when I first implemented these things.

Reply:
Yes, this was in some ways hacky but I wanted to avoid having the addional `In` wrapper as it seemed needlessly clunky, if we can achieve a more general approach I'm all for it but it doesn't seem straightforward.